### ISSUE TITLE
Add code generation for the advanced server interface

### ISSUE THREAD

**Issue by boosh** [grpc/grpc-java#339]
I asked on the mailing list (https://groups.google.com/forum/#!topic/grpc-io/1RiPC8XYP_U) whether a callback existed to notify the server that a client had disconnected so it could perform some cleanup actions. Here's the answer I received from Eric Anderson:

Because it is a unary request, there isn't a callback that notifies you of client cancellation (any error is considered a cancellation to the server). The next time you call onValue though, an exception will be thrown notifying you of the cancellation. That is enough in some scenarios.

The "advanced" interface provides such notification though. It has onHalfClose (which is when onComplete is normally called), onComplete (when the RPC gracefully completed), and onCancel (if there was an error or if the client cancelled). We don't have code generation for the advanced interface though, so you would need to implement ServerCallHandler for each method in the service, create ServerMethodDefinitions and add them to a ServerServiceDefinition, which is what you pass to the ServerBuilder's addService. That would be verbose and annoying, but simple code. It would probably make sense for you to create an issue for us to add codegen for the advanced interface.

**Comment by ejona86** [comment]
I would still like to see this happen, but I haven't been able to get support from others.

**Comment by bobymicroby** [comment]
:+1 for the The "advanced" interface

**Comment by rsmilyanov** [comment]
Yep, there are use cases where the advanced interface is needed.
@ejona86, are there any plans to have such interface?

**Comment by ejona86** [comment]
Still not enough support from others on the team. Once we have #1010 figured out for server-side, then I think it will be clear whether it is necessary.

**Comment by accresse** [comment]
I have a similar need for this feature.  I am trying to write a grpc server that subscribes to an external messaging broker upon client request and streams the messages back to the client.  When the client disconnects, I would like to stop the subscription.  Although I see an exception printed in the logs indicating "An existing connection was forcibly closed by the remote host", I don't see a way to relate that to a particular connection and perform clean up operations.

**Comment by ejona86** [comment]
@accresse, you can use `Context.addListener` or `ServerCallStreamObserver.setOnCancelHandler` (cast provided `StreamObserver`).

**Comment by youurayy** [comment]
Has anyone got this to work? Neither of the callbacks set via `Context.current().addListener` nor `ServerCallStreamObserver.setOnCancelHandler` is being called on the server in case when the client closes the connection.
There's the `"An existing connection was forcibly closed by the remote host"` being thrown internally, but there's no way to catch it.

**Comment by arunkjn** [comment]
hi @ejona86 , 
I managed to use `Context.current().addListener` inside my grpc method implementation on the server. I am able to detect broken stream. I wanted to understand the behaviour of this. Will it fire cancelled only inside the grpc method impl I am using it in? Also how does it compare to using `ServerCallStreamObserver.setOnCancelHandler`? I was not able to find sufficient documentation on `ServerCallStreamObserver.setOnCancelHandler ` could you please share some example or snippet if possible?

**Comment by zpencer** [comment]
@arunkjn Context is from grpc-context, and is actually a helper library that does not know anything about grpc itself. Each RPC runs inside an independent context, and gRPC will cancel the context when the request is complete.

`ServerCallStreamObserver.setOnCancelHandler` is for registering a cancellation callback that executes when a `ServerCall` is cancelled. 

So when a call is complete, gRPC first calls the cancellation handler of `ServerCallStreamObserver`, and then cancels the `Context`.

**Comment by arunkjn** [comment]
Got it. Thanks @zpencer

**Comment by ejona86** [comment]
@arunkjn, `ServerCallStreamObserver.setOnCancelHandler` is very similar to `Context.current().addListener`. They're mostly interchangeable, but `ServerCallStreamObserver`'s callback is serialized with the other callbacks while `Context`'s is not. That means that with `ServerCallStreamObserver` you may not need to synchronize your code, but it also means that blocking in any callback (e.g., `onNext()`) will prevent you from receiving the notification.
